lib/active_merchant/billing/gateways/ixopay.rb

```diff
@@ -67,55 +54,193 @@ def scrub(transcript)
+          xml.tag! 'expirationMonth', format(payment_method.month, :two_digits)
+          xml.tag! 'expirationYear', format(payment_method.year, :four_digits)
+        end
+      end
+
+      def add_debit(xml, money, options)
+        currency = options[:currency] || currency(money)
+        description = options[:description].blank? ? 'Purchase' : options[:description]
+
+        xml.tag! 'debit' do
+          xml.tag! 'transactionId', new_transaction_id
+          add_customer_data(xml, options)
+          xml.tag! 'amount', money
+          xml.tag! 'currency', currency
+          xml.tag! 'description', description
+          xml.tag! 'callbackUrl', options[:callback_url] || 'http://example.com'
+        end
       end
 
-      def add_address(post, creditcard, options)
+      def add_customer_data(xml, options)
+        # Ixopay returns an error if the elements are not added in the order used here.
+        xml.tag! 'customer' do
+          add_billing_address(xml,  options[:billing_address])  if options[:billing_address]
+          add_shipping_address(xml, options[:shipping_address]) if options[:shipping_address]
+
+          if options.dig(:billing_address, :company)
+            xml.tag! 'company', options[:billing_address][:company]
+          end
+
+          xml.tag! 'email', options[:email]
+          xml.tag! 'ipAddress', options[:ip] || '127.0.0.1'
+        end
       end
 
-      def add_invoice(post, money, options)
-        post[:amount] = amount(money)
-        post[:currency] = (options[:currency] || currency(money))
+      def add_billing_address(xml, address)
+        if address[:name]
+          xml.tag! 'firstName', split_names(address[:name])[0]
+          xml.tag! 'lastName',  split_names(address[:name])[1]
+        end
+
+        xml.tag! 'billingAddress1', address[:address1]
+        xml.tag! 'billingAddress2', address[:address2]
+        xml.tag! 'billingCity',     address[:city]
+        xml.tag! 'billingPostcode', address[:zip]
+        xml.tag! 'billingState',    address[:state]
+        xml.tag! 'billingCountry',  address[:country]
+        xml.tag! 'billingPhone',    address[:phone]
       end
 
-      def add_payment(post, payment)
+      def add_shipping_address(xml, address)
+        if address[:name]
+          xml.tag! 'shippingFirstName', split_names(address[:name])[0]
+          xml.tag! 'shippingLastName',  split_names(address[:name])[1]
+        end
+
+        xml.tag! 'shippingCompany',   address[:company]
+        xml.tag! 'shippingAddress1',  address[:address1]
+        xml.tag! 'shippingAddress2',  address[:address2]
+        xml.tag! 'shippingCity',      address[:city]
+        xml.tag! 'shippingPostcode',  address[:zip]
+        xml.tag! 'shippingState',     address[:state]
+        xml.tag! 'shippingCountry',   address[:country]
+        xml.tag! 'shippingPhone',     address[:phone]
       end
 
-      def parse(body)
-        {}
+      def new_transaction_id
```

Comment:
It doesn't look like this is used more than once. Why was it pulled out into a new method?